File: EmployeeOverview/RMA2_Roots/DBRoot.vb

Imports System.Data.Common
Imports System.Threading
Imports System.Text


'--------------------------------------------------------------------------------------------------------------------
'       ConnectionPool class
'
' usage for a non-parameterized ConnectionPool:
'    dummy = New ConnectionPool(Of NpgsqlConnection)("BankIO", "Server=localhost;Port=5432;User Id=rma;Password=rma;Database=rma_bankio;")
' This call registers a connection pool named 'BankIn'. The pooled connections use the given connection string.
'
' usage for a parameterized ConnectionPool:
' If there are many similar databases on the same server, it makes sense to have one 'configurable' ConnectionPool
' for all connections. A good example is SQL-Ledger; every client gets a set of (identical) tables in his own
' database. Here is the connection string:
'    dummy = New ConnectionPool(Of NpgsqlConnection)("SQLLedger", "Server=localhost;Port=5432;User Id=sql-ledger;Password=De99elria;Database={0}", True)
' Note the additional parameter True, and the "Database={0}" part in the string. When you get a connection from this pool,
' you have to specify a connParameter which replaces the {0} in the connection string.


Public Class ConnectionPool(Of connType As {DbConnection, New})

    Private name As String
    Private connStr As String
    Private configurable As Boolean
    Private pool As New Dictionary(Of String, List(Of connType))
    Private initialThreadID As Integer

    Sub New(ByVal poolname As String, ByVal connStr As String, Optional ByRef configurable As Boolean = False)
        Me.name = poolname
        Me.connStr = connStr
        Me.configurable = configurable

        ' store thread ID
        initialThreadID = Thread.CurrentThread.ManagedThreadId
    End Sub


    Public Function GetConnection(Optional ByVal connParameter As String = Nothing) As connType
        If configurable And connParameter Is Nothing Then
            Throw New ApplicationException("DBPool: Must specify connParameter for a configurable ConnectionPool.")
        ElseIf Not configurable Then
            If connParameter IsNot Nothing Then
                Throw New ApplicationException("DBPool: Can't specify connParameter for a non-configurable ConnectionPool")
            Else
                connParameter = ""
            End If
        End If
        connParameter = connParameter.Trim.ToLower()

        Dim conn As connType = Nothing
        Dim nonPooled = (initialThreadID <> Thread.CurrentThread.ManagedThreadId)

        If Not nonPooled Then
            If Not pool.ContainsKey(connParameter) Then
                ' create pool list, if it doesn't exist
                pool(connParameter) = New List(Of connType)
            Else
                ' pool exists - look for a available connection
                For Each testconn In pool(connParameter)
                    If (testconn.State = ConnectionState.Closed Or testconn.State = ConnectionState.Broken) Then
                        conn = testconn
                        Exit For
                    End If
                Next
            End If
        End If

        If conn Is Nothing Then
            ' no or no available connection - create a new one and store in pool
            conn = New connType
            If configurable Then
                conn.ConnectionString = String.Format(connStr, connParameter)
            Else
                conn.ConnectionString = connStr
            End If

            If Not nonPooled Then
                pool(connParameter).Add(conn)
            End If
        End If

        ' attempt to open the connection.. prepare for errors
        Try
            conn.Open()
        Catch ex As Exception
            Dim msg As String = "DBPool.GetConnection(Pool = " & name
            If configurable Then
                msg &= "." & connParameter
            End If
            msg &= ")" & vbLf & vbLf & String.Format("Die Datenbank '{0}' ist nicht erreichbar.", name)
            Throw New ApplicationException(msg)
        End Try
        Return conn
    End Function

    Public Sub EmptyPool(Optional ByVal connParameter As String = Nothing)
        If configurable And connParameter Is Nothing Then
            Throw New ApplicationException("DBPool: Must specify connParameter for a configurable ConnectionPool.")
        ElseIf Not configurable Then
            If connParameter IsNot Nothing Then
                Throw New ApplicationException("DBPool: Can't specify connParameter for a non-configurable ConnectionPool")
            Else
                connParameter = ""
            End If
        End If
        connParameter = connParameter.Trim.ToLower()

        If pool.ContainsKey(connParameter) Then
            ' pool exists - look for a available connection
            For i = pool(connParameter).Count - 1 To 0 Step -1
                Dim testconn = pool(connParameter)(i)
                If (testconn.State = ConnectionState.Closed Or testconn.State = ConnectionState.Broken) Then
                    pool(connParameter).RemoveAt(i)
                End If
            Next
        End If
    End Sub


    Public Shared Function GetCommand(ByRef conn As DbConnection, ByVal sql As String)
        Dim cmd As DbCommand = conn.CreateCommand
        cmd.CommandText = sql
        Return cmd
    End Function



    '--------------------------------------------------------------------------------------------------------------------
    '       Command executor
    '
    ' use this for INSERT, DELETE, UPDATE

    Public Function SQLExec(ByVal sql As String, Optional ByVal connParameter As String = Nothing) As Integer
        ' executes the given command(s) using the given connection or a connection of the specified pool
        ' & returns the number of affected rows.
        Dim conn = GetConnection(connParameter)
        Dim cmd = GetCommand(conn, sql)

        Dim affected As Integer = 0
        Try
            affected = cmd.ExecuteNonQuery()
        Finally
            conn.Close()
        End Try

        Return affected
    End Function



    '--------------------------------------------------------------------------------------------------------------------
    '       Data getters
    '

    Public Function SQL2O(ByVal sql As String, Optional ByVal connParameter As String = Nothing) As Object
        ' gets the requested data using the given connection or a connection of the specified pool
        ' & returns a single Object. use this for count(*) and such
        Dim conn = GetConnection(connParameter)
        Dim cmd = GetCommand(conn, sql)

        Dim result As Object = Nothing
        Try
            Dim dr = cmd.ExecuteReader()

            ' get data rows
            If dr.Read() Then
                result = dr.GetValue(0)
            End If
            dr.Close()

        Catch ex As Exception
        Finally
            conn.Close()
        End Try

        Return result
    End Function


    Public Function SQL2LO(ByVal sql As String, Optional ByVal connParameter As String = Nothing) As List(Of Object)
        ' gets the requested data using the given connection or a connection of the specified pool
        ' & returns a List of Object.
        ' If the query returns more than one column, only the first one is returned; the rest is ignored.
        Dim conn = GetConnection(connParameter)
        Dim cmd = GetCommand(conn, sql)

        Dim resultSet As New List(Of Object)
        Try
            Dim dr = cmd.ExecuteReader()

            ' get data rows
            While (dr.Read())
                Dim rec As Object = New Object
                rec = dr.GetValue(0)
                resultSet.Add(rec)
            End While
            dr.Close()

        Catch ex As Exception
        Finally
            conn.Close()
        End Try

        Return resultSet
    End Function


    Public Function SQL2LAO(ByVal sql As String, Optional ByVal connParameter As String = Nothing) As List(Of Object())
        ' gets the requested data using the given connection or a connection of the specified pool
        ' & returns a List of Object()
        Dim conn = GetConnection(connParameter)
        Dim cmd = GetCommand(conn, sql)

        Dim resultSet As New List(Of Object())
        Try
            Try
                Dim dr = cmd.ExecuteReader()
                Dim items As Integer = dr.FieldCount

                ' get data rows
                While (dr.Read())
                    Dim rec As Object() = New Object(items - 1) {}
                    dr.GetValues(rec)
                    resultSet.Add(rec)
                End While
                dr.Close()
            Finally
                conn.Close()
            End Try

        Catch ex As Exception
        End Try

        Return resultSet
    End Function



    Public Function SQL2LD(ByVal sql As String, Optional ByVal connParameter As String = Nothing) As List(Of RawDataRecord)
        ' gets the requested data using the given connection or a connection of the specified pool
        ' & returns a List of RawDataRecord
        Dim conn = GetConnection(connParameter)
        Dim cmd = GetCommand(conn, sql)
        Dim resultList As New List(Of RawDataRecord)
        Try
            Try
                Dim dr = cmd.ExecuteReader()
                Dim items As Integer = dr.FieldCount

                ' get column names
                Dim cNames As String() = New String(items - 1) {}
                For i As Integer = 0 To items - 1
                    cNames(i) = dr.GetName(i)
                Next

                ' get data rows
                Dim rec As Object() = New Object(items - 1) {}
                While (dr.Read())
                    Try
                        dr.GetValues(rec)

                        ' .. and fill into Directory
                        Dim rdr As New RawDataRecord
                        For i As Integer = 0 To items - 1
                            rdr(cNames(i)) = rec(i)
                        Next

                        resultList.Add(rdr)
                    Catch ex As Exception
                    End Try
                End While
                dr.Close()
            Finally
                conn.Close()
            End Try

        Catch ex As Exception
        End Try

        Return resultList
    End Function

End Class


' This is what SQL2LD(...) returns
Public Class RawDataRecord
    Inherits Dictionary(Of String, Object)

    '--------------------------------------------------------------------------------------------------------------------
    '       Data translators
    '
    ' these members are used to fill DB data into VB data objects. Empty fields,
    ' special contents and formatting can be processed/corrected here

    Public Function AsString(ByVal key As String, Optional ByVal trim As Boolean = True, Optional ByVal nilAsEmpty As Boolean = False) As String
        Dim value = Me(key)
        If IsDBNull(value) Or value Is Nothing Then
            If nilAsEmpty Then
                Return ""
            Else
                Return Nothing
            End If
        ElseIf Not TypeOf value Is String Then
            value = value.ToString()
        End If

        If trim Then
            Return value.Trim()
        End If
        Return value
    End Function

    Public Function AsChar(ByVal key As String) As Char
        Dim value = Me(key)
        If TypeOf value Is Char Then
            Return value
        ElseIf TypeOf value Is String AndAlso value.ToString.Length > 0 Then
            Return value.ToString(0)
        End If
        Return ""
    End Function

    Public Function AsBoolean(ByVal key As String) As Boolean
        Dim value = Me(key)
        If TypeOf value Is Boolean Then
            Return value
        ElseIf IsDBNull(value) Or value Is Nothing Then
            Return False
        ElseIf value Like "*1*" Then
            Return True
        Else
            Return False
        End If
    End Function

    Public Function AsBooleanNIL(ByVal key As String) As Boolean?
        Dim value = Me(key)
        If TypeOf value Is Boolean Then
            Return value
        ElseIf IsDBNull(value) Or value Is Nothing Then
            Return Nothing
        ElseIf value Like "*1*" Then
            Return True
        Else
            Return False
        End If
    End Function

    Public Function AsInteger(ByVal key As String, Optional ByVal nullAsMinInt As Boolean = False) As Integer
        Dim value = Me(key)
        If IsDBNull(value) Or value Is Nothing Then
            Return If(nullAsMinInt, Integer.MinValue, 0)
        End If
        Return value
    End Function

    Public Function AsIntegerNIL(ByVal key As String, Optional ByVal nullAsMinInt As Boolean = False) As Integer?
        Dim value = Me(key)
        If IsDBNull(value) Or value Is Nothing Then
            Return Nothing
        End If
        Return value
    End Function

    Public Function AsDouble(ByVal key As String, Optional ByVal nilAsZero As Boolean = True) As Double
        Dim value = Me(key)
        If IsDBNull(value) OrElse value Is Nothing Then
            If nilAsZero Then
                Return 0
            Else
                Return Nothing
            End If
        End If

        Dim thisDouble As Double
        Try
            thisDouble = value
        Catch ex As Exception
            If nilAsZero Then
                Return 0
            Else
                Return Nothing
            End If
        End Try
        Return thisDouble
    End Function

    Public Function AsDoubleNIL(ByVal key As String, Optional ByVal nilAsZero As Boolean = True) As Double?
        Dim value = Me(key)
        If IsDBNull(value) OrElse value Is Nothing Then
            Return Nothing
        End If

        Dim thisDouble As Double
        Try
            thisDouble = value
        Catch ex As Exception
            Return Nothing
        End Try
        Return thisDouble
    End Function

    Public Function AsDate(ByVal key As String) As Date
        Dim value = Me(key)
        If IsDBNull(value) Or value Is Nothing Then
            Return Nothing
        End If

        Dim thisDate As Date
        Try
            thisDate = value
        Catch ex As Exception
            Return Nothing
        End Try
        Return thisDate
    End Function

    Public Function AsDateNIL(ByVal key As String) As Date?
        Dim value = Me(key)
        If IsDBNull(value) Or value Is Nothing Then
            Return Nothing
        End If

        Dim thisDate As Date?
        Try
            thisDate = value
        Catch ex As Exception
            Return Nothing
        End Try
        Return thisDate
    End Function

    Public Function AsXML2Dictionary(ByVal key As String) As DictionaryWithDefault(Of String, Object)
        ' parses the header XML field & the transaction XML field and returns all properties in a new dictionary.
        Dim xml = Me(key)
        Dim value As New DictionaryWithDefault(Of String, Object)

        Try
            Dim xmlElem = XElement.Parse(xml)
            For Each attr In xmlElem.Attributes
                value.Add(attr.Name.ToString, attr.Value)
            Next
        Catch ex As Exception
        End Try

        Return value
    End Function

    Public Overrides Function ToString() As String
        Dim result = New StringBuilder()
        For Each key In Keys.OrderBy(Function(_k) _k)
            result.Append(key & Me(key))
        Next
        Return result.ToString
    End Function
End Class

File: EmployeeOverview/WPF/WPF_Roots/DBRoot.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Xml.Linq;

namespace WPF_Roots
{

    public class DBException : Exception
    {
        public DBException(string msg) : base(msg) { }
    }

    public abstract class ConnectionPool 
    {
        public abstract DbConnection GetConnection();
    };

    public class ConnectionPool<ConnectionType> : ConnectionPool 
           where ConnectionType: DbConnection, new()
    {
        private string name;
        private string connectionStr;
        private List<ConnectionType> pool = new List<ConnectionType>();
        private int initialThreadID;


        public ConnectionPool(string poolName, string connectionString)
        {
            this.name = poolName;
            this.connectionStr = connectionString;
            initialThreadID = Thread.CurrentThread.ManagedThreadId;
        }


        public override DbConnection GetConnection()
        {
            var isOnMainThread = (initialThreadID == Thread.CurrentThread.ManagedThreadId);
            ConnectionType conn = null;

            if (isOnMainThread)
                conn = pool.Find(_c => (_c.State == System.Data.ConnectionState.Closed || _c.State == System.Data.ConnectionState.Broken));

            if (conn == null)
            {
                conn = new ConnectionType();
                conn.ConnectionString = connectionStr;

                if (isOnMainThread)
                    pool.Add(conn);
            }

            // try to open the connection
            try
            {
                conn.Open();
            }
            catch (Exception ex)
            {
                var msg = string.Format("ConnectionPool.GetConnection():\nFehler beim Öffnen der DB-Verbindung (pool='{0}'):\n{1}", name, ex.Message);
                throw new DBException(msg);
            }

            //
            return conn;
        }


        public DbCommand GetCommand(DbConnection dbc, string sql)
        {
            var cmd = dbc.CreateCommand();
            cmd.CommandText = sql;
            return cmd;
        }



        /// SQLExec
        /// command execution - number of affected rows is returned
        /// 
        public int SQLExec(string sql, DbConnection dbc = null)
        {
            var useOwnConnection = (dbc == null);
            dbc = dbc ?? GetConnection();

            var cmd = GetCommand(dbc, sql);
            int affectedRows = 0;

            try
            {
                affectedRows = cmd.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                var msg = string.Format("ConnectionPool.SQLExec():\nFehler bei SQL-Abfrage (pool='{0}'):\n{1}\nSQL: {2}", name, ex.Message, sql);
                throw new DBException(msg);
            }
            finally
            {
                if (useOwnConnection)
                    dbc.Close();
            }

            return affectedRows;
        }

        /// SQL2O
        /// executes the given sql command & returns exactly 1 value.
        /// for example: 'insert into .. values (..) returning id'
        /// 
        public object SQL2O(string sql, DbConnection dbc = null)
        {
            var useOwnConnection = (dbc == null);
            dbc = dbc ?? GetConnection();

            var cmd = GetCommand(dbc, sql);
            DbDataReader dr = null;

            object result = null;
            try
            {
                dr = cmd.ExecuteReader();
                if (dr.Read())
                    result = dr.GetValue(0);
            }
            catch (Exception ex)
            {
                var msg = string.Format("ConnectionPool.SQL2O():\nFehler bei SQL-Abfrage (pool='{0}'):\n{1}\nSQL: {2}", name, ex.Message, sql);
                throw new DBException(msg);
            }
            finally
            {
                if (dr != null && !dr.IsClosed)
                    dr.Close();
                if (useOwnConnection)
                    dbc.Close();
            }

            return result;
        }


        /// SQL2LO
        /// executes the given sql command & returns a list of objects.
        /// 
        public List<object> SQL2LO(string sql, DbConnection dbc = null)
        {
            var useOwnConnection = (dbc == null);
            dbc = dbc ?? GetConnection();

            var cmd = GetCommand(dbc, sql);
            DbDataReader dr = null;

            var result = new List<Object>();
            try
            {
                dr = cmd.ExecuteReader();
                while (dr.Read())
                    result.Add(dr.GetValue(0));
            }
            catch (Exception ex)
            {
                var msg = string.Format("ConnectionPool.SQL2LO():\nFehler bei SQL-Abfrage (pool='{0}'):\n{1}\nSQL: {2}", name, ex.Message, sql);
                throw new DBException(msg);
            }
            finally
            {
                if (dr != null && !dr.IsClosed)
                    dr.Close();
                if (useOwnConnection)
                    dbc.Close();
            }

            return result;
        }


        /// SQL2LAO
        /// executes the given sql command & returns a list of records, coded as object[].
        /// 
        public List<object[]> SQL2LAO(string sql, DbConnection dbc = null)
        {
            var useOwnConnection = (dbc == null);
            dbc = dbc ?? GetConnection();

            var cmd = GetCommand(dbc, sql);
            DbDataReader dr = null;

            var result = new List<Object[]>();
            try
            {
                dr = cmd.ExecuteReader();
                var items = dr.FieldCount;

                while (dr.Read())
                {
                    var rec = new object[items];
                    dr.GetValues(rec);
                    result.Add(rec);
                }
            }
            catch (Exception ex)
            {
                var msg = string.Format("ConnectionPool.SQL2LAO():\nFehler bei SQL-Abfrage (pool='{0}'):\n{1}\nSQL: {2}", name, ex.Message, sql);
                throw new DBException(msg);
            }
            finally
            {
                if (dr != null && !dr.IsClosed)
                    dr.Close();
                if (useOwnConnection)
                    dbc.Close();
            }

            return result;
        }


        /// SQL2RD
        /// executes the given sql command & returns a List of RawDataRecords.
        /// 
        public List<RawDataRecord> SQL2RD(string sql, DbConnection dbc = null)
        {
            var useOwnConnection = (dbc == null);
            dbc = dbc ?? GetConnection();

            var cmd = GetCommand(dbc, sql);
            DbDataReader dr = null;
            object[] rowValues;

            var result = new List<RawDataRecord>();
            try
            {
                dr = cmd.ExecuteReader();

                // get column names
                var columnCount = dr.FieldCount;
                var columnNames = new string[columnCount];
                for (var i = 0; i < columnCount; i++)
                    columnNames[i] = dr.GetName(i);

                // get data rows
                rowValues = new object[columnCount];
                while (dr.Read())
                {
                    dr.GetValues(rowValues);
                    var rdr = new RawDataRecord(columnNames, rowValues);
                    result.Add(rdr);
                }
            }
            catch (Exception ex)
            {
                var msg = string.Format("ConnectionPool.SQL2RD():\nFehler bei SQL-Abfrage (pool='{0}'):\n{1}\nSQL: {2}", name, ex.Message, sql);
                throw new DBException(msg);
            }
            finally
            {
                if (dr != null && !dr.IsClosed)
                    dr.Close();
                if (useOwnConnection)
                    dbc.Close();
            }

            return result;
        }


        /// LoadDBItems
        /// loads a list of an arbitrary type T using reflection - 
        /// the fields of T must be properties & must have the same name as the fields in the DB table
        /// 
        public List<T> LoadDBItems<T>(String sql) where T : new()
        {
            var result = new List<T>();

            var conn = GetConnection();
            var cmd = GetCommand(conn, sql);
            DbDataReader dr = null;

            var recordCounter = 0;
            try
            {
                dr = cmd.ExecuteReader();
                int fieldCount = dr.FieldCount;

                // load column names
                var columnNames = new List<String>();
                for (int i = 0; i < fieldCount; i++)
                    columnNames.Add(dr.GetName(i));

                var rec = new Object[fieldCount];
                while (dr.Read())
                {
                    recordCounter++;
                    dr.GetValues(rec);
                    var jo = new JObject();
                    for (int i = 0; i < fieldCount; i++)
                    {
                        var value = rec[i];
                        if (value != null && value != DBNull.Value)
                            jo.Add(columnNames[i], new JValue(value));
                    }

                    var newT = new T();
                    JsonConvert.PopulateObject(jo.ToString(), newT);
                    result.Add(newT);
                }
            }
            catch (Exception ex)
            {
                var msg = string.Format("ConnectionPool.LoadDBItems():\nFehler bei SQL-Abfrage (pool='{0}'):\n{1}\nSQL: {2}", name, ex.Message, sql);
                throw new DBException(msg);
            }
            finally
            {
                if (dr != null && !dr.IsClosed)
                    dr.Close();
                conn.Close();
            }

            return result;
        }

    }


    public class RawDataRecord : Dictionary<string, object>
    {

        public RawDataRecord() : base() { }

        public RawDataRecord(string[] columnNames, object[] rowValues) : base()
        {
            var fieldCount = columnNames.Length;
            if (fieldCount != rowValues.Length)
                throw new DBException("DBRoot.RawDataRecord: Initialisations-Fehler.");

            for (var i = 0; i < fieldCount; i++)
                this[columnNames[i]] = rowValues[i];
        }

        public override string ToString()
        {
            var orderedParts = new List<string>();
            foreach (var columnName in this.Keys.OrderBy(k => k))
                orderedParts.Add(string.Format("{0}:{1}", columnName, this[columnName]));

            return string.Join("\n", orderedParts.ToArray());
        }


        // data getters

        public string AsString(string key, bool trim = true, bool nullAsEmpty = false)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value == null || DBNull.Value.Equals(value))
                return nullAsEmpty ? String.Empty : null;

            if (trim)
                return value.ToString().Trim();

            return value.ToString();
        }


        public char AsChar(string key, char valueIfNull = '?')
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value is char)
                return (char)value;
            if (value == null || DBNull.Value.Equals(value))
                return valueIfNull;

            var result = value.ToString();
            if (result.Length > 0)
                return result[0];

            return valueIfNull;
        }


        public bool AsBoolean(string key, bool valueIfNull = false)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value is Boolean)
                return (bool)value;

            else if (value == null || DBNull.Value.Equals(value))
                return valueIfNull;

            bool result = false;
            if (bool.TryParse(value.ToString(), out result))
                return result;

            return false;
        }

        public bool? AsBooleanNIL(string key)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value is Boolean)
                return (bool)value;

            else if (value == null || DBNull.Value.Equals(value))
                return null;

            bool result = false;
            if (bool.TryParse(value.ToString(), out result))
                return result;

            return false;
        }


        public int AsInteger(string key, bool nullAsMinValue = false)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value == null || DBNull.Value.Equals(value))
                return nullAsMinValue ? int.MinValue : 0;

            int result = 0;
            if (int.TryParse(value.ToString(), out result))
                return result;

            throw new DBException(string.Format("DBRoot.RawDataRecord.AsInteger: '{0}' kann nicht als Integer dargestellt werden (key = '{1}').", value.ToString(), key));
        }

        public int? AsIntegerNIL(string key)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value == null || DBNull.Value.Equals(value))
                return null;

            int result = 0;
            if (int.TryParse(value.ToString(), out result))
                return result;

            throw new DBException(string.Format("DBRoot.RawDataRecord.AsIntegerNIL: '{0}' kann nicht als Integer dargestellt werden (key = '{1}').", value.ToString(), key));
        }


        public double AsDouble(string key)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value == null || DBNull.Value.Equals(value))
                return 0.0;

            double result = 0.0;
            if (double.TryParse(value.ToString(), out result))
                return result;

            throw new DBException(string.Format("DBRoot.RawDataRecord.AsDouble: '{0}' kann nicht als Double dargestellt werden (key = '{1}').", value.ToString(), key));
        }


        public double? AsDoubleNIL(string key)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value == null || DBNull.Value.Equals(value))
                return null;

            double result = 0.0;
            if (double.TryParse(value.ToString(), out result))
                return result;

            return null;
        }


        public DateTime AsDate(string key)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            DateTime result = DateTime.MinValue;

            if (value == null || DBNull.Value.Equals(value))
                return result;

            if (DateTime.TryParse(value.ToString(), out result))
                return result;

            throw new DBException(string.Format("DBRoot.RawDataRecord.AsDate: '{0}' kann nicht als DateTime dargestellt werden (key = '{1}').", value.ToString(), key));
        }


        public DateTime? AsDateNIL(string key)
        {
            if (!ContainsKey(key))
                throw new DBException(string.Format("DBRoot.RawDataRecord: Unbekannter Key {0}.", key == null ? "<NULL>" : "'" + key + "'"));

            var value = this[key];
            if (value == null || DBNull.Value.Equals(value))
                return null;

            DateTime result = DateTime.MinValue;
            if (DateTime.TryParse(value.ToString(), out result))
                return result;

            return null;
        }

    }

}
